src/ripple/nodestore/impl/DatabaseImp.h

```diff
@@ -92,17 +91,15 @@ class DatabaseImp
             fdlimit_ = m_backend->fdlimit();
     }
 
-    ~DatabaseImp ()
+    ~DatabaseImp () override
     {
-        {
-            std::lock_guard <std::mutex> lock (m_readLock);
-            m_readShut = true;
-            m_readCondVar.notify_all ();
-            m_readGenCondVar.notify_all ();
-        }
-
-        for (auto& e : m_readThreads)
-            e.join();
+        // NOTE!
+        // Any derived class should call the stopThreads() method in its
+        // destructor.  Otherwise, occasionally, the derived class may
+        // crash during shutdown when its members are accessed by one of
+        // these threads after the derived class is destroyed but before
+        // this base class is destroyed.
+        stopThreads();
```

Comment:
Yeah, I thought about this too and didn't come up with anything that didn't require a bunch of restructuring.  I didn't think it was worth a lot of effort since it feels like we're unlikely to derive yet another kind of Database from `DatabaseImp`.